IndexedSeqOptimized: tailrec bugs

```
/scala-scala/src/library/scala/collection/IndexedSeqOptimized.scala:58: error: Cannot rewrite recursive call: it changes type of 'this' on a polymorphic recursive call
    if (start == end) z
    ^
./scala-scala/src/library/scala/collection/IndexedSeqOptimized.scala:57: error: TailRec optimisation not applicable, method not tail recursive
  private def foldl[B](start: Int, end: Int, z: B, op: (B, A) => B): B =
              ^
./scala-scala/src/library/scala/collection/IndexedSeqOptimized.scala:63: error: Cannot rewrite recursive call: it changes type of 'this' on a polymorphic recursive call
    if (start == end) z
    ^
./scala-scala/src/library/scala/collection/IndexedSeqOptimized.scala:62: error: TailRec optimisation not applicable, method not tail recursive
  private def foldr[B](start: Int, end: Int, z: B, op: (A, B) => B): B =
              ^
```

--- a/src/dotty/tools/dotc/Compiler.scala
+++ b/src/dotty/tools/dotc/Compiler.scala
@@ -52,6 +52,7 @@ class Compiler {
       List(new PatternMatcher,
            new ExplicitOuter,
            new ExplicitSelf,
+           new CrossCastAnd,
            new Splitter),
       List(new VCInlineMethods,
            new SeqLiterals,

--- a/src/dotty/tools/dotc/transform/CrossCastAnd.scala
+++ b/src/dotty/tools/dotc/transform/CrossCastAnd.scala
@@ -20,29 +20,11 @@ class CrossCastAnd extends MiniPhaseTransform { thisTransform =>
   override def phaseName: String = "crossCast"
 
   override def transformSelect(tree: tpd.Select)(implicit ctx: Context, info: TransformerInfo): tpd.Tree = {
-    if (tree.symbol.is(Flags.Private)) {
-      // all private member selections have a symbol
-      tree.qualifier.tpe.widen match {
-        case t @ AndType(l, _) =>
 
-        // find a component of and type that owns the symbol
-        def findType(tp: Type): Type = {
-            tp match {
-              case AndType(l, r) =>
-                findType(l).orElse(findType(r))
-              case t =>
-                if (t.decl(tree.symbol.name).suchThat(_ == tree.symbol).exists)
-                  t
-                else NoType
-            }
-          }
-
-          val tp = findType(t)
-          if (l eq tp) tree
-          else tree.qualifier.asInstance(AndType(tp, tree.qualifier.tpe)).select(tree.symbol)
-        case _ => tree
-      }
-    }
-      else tree
+    lazy val qtype = tree.qualifier.tpe.widen
+    val sym = tree.symbol
+    if (sym.is(Flags.Private) && qtype.typeSymbol != sym.owner)
+      cpy.Select(tree)(tree.qualifier.asInstance(AndType(qtype.baseTypeWithArgs(sym.owner), tree.qualifier.tpe)), tree.name)
+    else tree
   }
 }

--- a/src/dotty/tools/dotc/transform/CrossCastAnd.scala
+++ b/src/dotty/tools/dotc/transform/CrossCastAnd.scala
@@ -1,0 +1,48 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Flags
+import dotty.tools.dotc.core.Types.{NoType, Type, AndType}
+import dotty.tools.dotc.transform.TreeTransforms._
+import tpd._
+
+import scala.collection.mutable.ListBuffer
+
+
+/**
+ * This transform makes sure that all private member selections from
+ * AndTypes are performed from the first component of AndType.
+ * This is needed for correctness of erasure. See `tests/run/PrivateAnd.scala`
+ */
+class CrossCastAnd extends MiniPhaseTransform { thisTransform =>
+
+  override def phaseName: String = "crossCast"
+
+  override def transformSelect(tree: tpd.Select)(implicit ctx: Context, info: TransformerInfo): tpd.Tree = {
+    if (tree.symbol.is(Flags.Private)) {
+      // all private member selections have a symbol
+      tree.qualifier.tpe.widen match {
+        case t @ AndType(l, _) =>
+
+        // find a component of and type that owns the symbol
+        def findType(tp: Type): Type = {
+            tp match {
+              case AndType(l, r) =>
+                findType(l).orElse(findType(r))
+              case t =>
+                if (t.decl(tree.symbol.name).suchThat(_ == tree.symbol).exists)
+                  t
+                else NoType
+            }
+          }
+
+          val tp = findType(t)
+          if (l eq tp) tree
+          else tree.qualifier.asInstance(AndType(tp, t)).select(tree.symbol)
+        case _ => tree
+      }
+    }
+      else tree
+  }
+}

--- a/src/dotty/tools/dotc/transform/CrossCastAnd.scala
+++ b/src/dotty/tools/dotc/transform/CrossCastAnd.scala
@@ -39,7 +39,7 @@ class CrossCastAnd extends MiniPhaseTransform { thisTransform =>
 
           val tp = findType(t)
           if (l eq tp) tree
-          else tree.qualifier.asInstance(AndType(tp, t)).select(tree.symbol)
+          else tree.qualifier.asInstance(AndType(tp, tree.qualifier.tpe)).select(tree.symbol)
         case _ => tree
       }
     }

--- a/src/dotty/tools/dotc/transform/TailRec.scala
+++ b/src/dotty/tools/dotc/transform/TailRec.scala
@@ -185,7 +185,7 @@ class TailRec extends MiniPhaseTransform with DenotTransformer with FullParamete
 
         val receiverIsSame = enclosingClass.typeRef.widenDealias =:= recvWiden
         val receiverIsSuper = (method.name eq sym) && enclosingClass.typeRef.widen <:< recvWiden
-        val receiverIsThis = recv.tpe =:= thisType
+        val receiverIsThis = recv.tpe =:= thisType || recv.tpe.widen =:= thisType
 
         val isRecursiveCall = (method eq sym)
 

--- a/tests/pos/ErasureAnd.scala
+++ b/tests/pos/ErasureAnd.scala
@@ -1,0 +1,11 @@
+import scala.annotation.tailrec
+trait A { self: B =>
+  @tailrec
+  private def foo(arg1: Int, arg2: Int): Int = {
+    def bar = this.foo(arg1, arg2)
+    foo(arg1, arg2)
+  }
+  def foo(arg: Int) = arg
+}
+
+class B extends A{}

--- a/tests/run/PrivateAnd.scala
+++ b/tests/run/PrivateAnd.scala
@@ -1,0 +1,17 @@
+class A{
+  private def foo = 1
+  def callsFoo1(other: A & B): Int = other.foo
+  def callsFoo2(other: B & A): Int = other.foo
+}
+
+trait B {
+  def foo(i: Int) = i
+}
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    val a = new A with B
+    a.callsFoo1(a)
+    a.callsFoo2(a)	
+  }
+}

